Added valid_move? for bishop

guillaumebihet opened the pull request:
Added move_to? method for bishop and tests

guillaumebihet committed:
Added valid_move? for bishop

guillaumebihet committed:
Added queen move_to? and tests

ilyakrasnov commented on app/models/bishop.rb:
```diff
@@ -7,4 +7,22 @@ def self.get_image(color)
+
+
+ def valid_move?(x,y)
+    if diagonal_move?(x,y) && is_on_board?(x,y)
```
It can be just one line `diagonal.. & is_on_board`.. will automatically return true/false :)

ilyakrasnov commented on app/models/queen.rb:
```diff
@@ -8,4 +8,30 @@ def self.get_image(color)
   end  
+
+ def valid_move?(x,y)
+    if (diagonal_move?(x,y) || vertical_move?(x,y)) && is_on_board?(x,y)
```
same as above in diagonal_move - can be one line..

ilyakrasnov commented on app/models/bishop.rb:
```diff
@@ -7,4 +7,22 @@ def self.get_image(color)
+    @current_x = self.position_x
+    @current_y = self.position_y
+
+    (@current_x-x).abs == (@current_y-y).abs
```
do we really need @current_x? it can also be: `(self.position_x - x)`

guillaumebihet replied:
Indeed that was pretty useless.. refactored, thanks!

ilyakrasnov reviewed (dismissed):
Great job on the tests. A few methods can be refactored to be much shorter (see comments)..

guillaumebihet committed:
moved is_obstructed test from controller to model, building new tests

ilyakrasnov commented on app/models/bishop.rb:
```diff
@@ -7,4 +7,19 @@ def self.get_image(color)
+
+
+ def valid_move?(x,y)
+     diagonal_move?(x,y) && is_on_board?(x,y) 
```
Awesome!

guillaumebihet replied:
thanks!

guillaumebihet committed:
refactored move_to? methods for queen, bishop, knight and king

guillaumebihet committed:
little refactoring forgotten for the queen move_ot?

guillaumebihet committed:
added the is_obstructed? check to valid_move? for bishop and queen

guillaumebihet commented:
Added is_obstructed? for valid_move? for queen and bishop.
Couldn't make it work for two hours, till Ilya pointed the finger on the populate_game! after_create action, that puts all the pieces on the board, so of course I had many obstructed squares without realising it. 
The tests could be done in the middle of the board to avoid getting obstructed by the other ones.
Instead, there is a way to cancel after_create call back, as such: 
Game.skip_callback(:create, :after, :populate_game!, raise: false)
Here is Ilya's link explaining it:
http://flipruby.com/skip-callbacks-on-factory-girl-and-rspec.html

Note the "raise: false" at the end. I found that in the following post to solve a problem I had with this method:
https://github.com/thoughtbot/factory_bot/issues/931
A kind of mismach between rails and factoryBot, but this solves the issue.

I also added this code to the future rook tests.
